pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go

```diff
@@ -0,0 +1,436 @@
+)
+
+type ReduceAggConfig struct {
+	// Configuration on how to aggregate one or more data points
+	Fields []AggregationField `mapstructure:"fields"  required:"true"`
+	// The top level field name that report data is put into
+	OutputFieldName string `mapstructure:"outputFieldName" json:"outputFieldName" default:"Reports"`
+	// The structure surrounding the report data that is put on to "OutputFieldName"
+	ReportFormat string `mapstructure:"reportFormat" json:"reportFormat" default:"map" jsonschema:"enum=map,enum=array"`
+}
+
+type AggregationField struct {
+	// The key to find a data point within the input data
+	// If omitted, the entire input will be used
+	InputKey string `mapstructure:"inputKey" json:"inputKey"`
+	// The key that the aggregated data is put under
+	// If omitted, the InputKey will be used
+	OutputKey string `mapstructure:"outputKey" json:"outputKey"`
+	// How the data set should be aggregated
+	Method string `mapstructure:"method" json:"method" jsonschema:"enum=median,enum=mode" required:"true"`
+	// An optional check to only report when the difference from the previous report exceeds a certain threshold.
+	// Can only be used when the field is of a numeric type: string, decimal, int64, big.Int, time.Time, float64
+	// If no deviation is provided on any field, there will always be a report once minimum observations are reached.
+	DeviationString string          `mapstructure:"deviation"  json:"deviation,omitempty"`
+	Deviation       decimal.Decimal `mapstructure:"-"  json:",omitempty"`
+	// The format of the deviation being provided
+	// * percent - a percentage deviation
+	// * absolute - an unsigned numeric difference
+	DeviationType string `mapstructure:"deviationType" json:"deviationType,omitempty" jsonschema:"enum=percent,enum=absolute,enum=none"`
+}
+
+type reduceAggregator struct {
+	config ReduceAggConfig
+}
+
+var _ types.Aggregator = (*reduceAggregator)(nil)
+
+func (a *reduceAggregator) Aggregate(lggr logger.Logger, previousOutcome *types.AggregationOutcome, observations map[ocrcommon.OracleID][]values.Value, f int) (*types.AggregationOutcome, error) {
+	if len(observations) < 2*f+1 {
+		return nil, fmt.Errorf("not enough observations, have %d want %d", len(observations), 2*f+1)
+	}
+
+	currentState, err := a.initializeCurrentState(lggr, previousOutcome)
+	if err != nil {
+		return nil, err
+	}
+
+	report := map[string]any{}
+	shouldReport := false
+
+	for _, field := range a.config.Fields {
+		vals := a.extractValues(lggr, observations, field.InputKey)
+
+		// only proceed if every field has reached the minimum number of observations
+		if len(vals) < 2*f+1 {
+			return nil, fmt.Errorf("not enough observations provided %s, have %d want %d", field.InputKey, len(vals), 2*f+1)
+		}
+
+		singleValue, err := reduce(field.Method, vals)
+		if err != nil {
+			return nil, fmt.Errorf("unable to reduce on method %s, err: %s", field.Method, err.Error())
+		}
+
+		if field.DeviationType != "none" {
```

Comment:
Can we put this in a constant, along with the other deviation types :)

Reply:
Yes!